### ISSUE TITLE
DnsCachedResolver deadlocks

### ISSUE THREAD

**Issue by graemerobertson** [Metaswitch/cpp-common#253]
Seen running the memento live tests. Whilst idle, Memento had all 50 of its worker threads in the state below, and as a result wasn't replying to HTTP requests. I think it's also been seen in Sprout. Rob is currently working on a fix for this - https://github.com/Metaswitch/cpp-common/pull/252.

```
#0  0x00007f6b009d489c in __lll_lock_wait () from /lib/x86_64-linux-gnu/libpthread.so.0
#1  0x00007f6b009d0080 in _L_lock_903 () from /lib/x86_64-linux-gnu/libpthread.so.0
#2  0x00007f6b009cff19 in pthread_mutex_lock () from /lib/x86_64-linux-gnu/libpthread.so.0
#3  0x0000000000485c33 in DnsCachedResolver::dns_query (this=0x209b4b0, domains=..., dnstype=1, results=...)
    at /var/lib/jenkins/workspace/memento/modules/cpp-common/src/dnscachedresolver.cpp:203
#4  0x0000000000486655 in DnsCachedResolver::dns_query (this=0x209b4b0, domain=..., dnstype=1)
    at /var/lib/jenkins/workspace/memento/modules/cpp-common/src/dnscachedresolver.cpp:190
#5  0x00000000004711fb in BaseResolver::a_resolve (this=0x209b560, hostname=..., af=<optimized out>, port=8888, transport=6,
    retries=5, targets=..., ttl=@0x7f6ac5fc9d9c: 0, trail=7458)
    at /var/lib/jenkins/workspace/memento/modules/cpp-common/src/baseresolver.cpp:345
#6  0x0000000000488b9b in HttpResolver::resolve (this=0x209b560, host=..., port=8888, max_targets=5, targets=..., trail=7458)
    at /var/lib/jenkins/workspace/memento/modules/cpp-common/src/httpresolver.cpp:85
#7  0x00000000004469d4 in HttpConnection::send_request (this=0x209b600, path=..., body=..., doc=..., username=..., trail=7458,
    method_str=..., curl=0x7f6ae00a7500) at /var/lib/jenkins/workspace/memento/modules/cpp-common/src/httpconnection.cpp:480
#8  0x0000000000442710 in HttpConnection::send_get (this=0x209b600, path=..., headers=..., response=..., username=...,
    trail=7458) at /var/lib/jenkins/workspace/memento/modules/cpp-common/src/httpconnection.cpp:424
#9  0x000000000044478f in HttpConnection::send_get (this=<optimized out>, path=..., response=..., username=...,
    trail=<optimized out>) at /var/lib/jenkins/workspace/memento/modules/cpp-common/src/httpconnection.cpp:410
#10 0x000000000044bb92 in HomesteadConnection::get_digest_and_parse (this=0x2090230, path=..., digest=..., realm=...,
    trail=7458) at homesteadconnection.cpp:111
#11 0x000000000044b4dc in HomesteadConnection::get_digest_data (this=0x2090230, private_user_identity=...,
    public_user_identity=..., digest=..., realm=..., trail=7458) at homesteadconnection.cpp:80
#12 0x000000000044cd11 in HTTPDigestAuthenticate::request_digest_and_store (this=0x7f6ae0057db0, www_auth_header=...,
    include_stale=false, response=<optimized out>) at httpdigestauthenticate.cpp:316
#13 0x000000000045012d in HTTPDigestAuthenticate::authenticate_request (this=0x7f6ae0057db0, impu=...,
    authorization_header=..., www_auth_header=..., method=..., trail=<optimized out>) at httpdigestauthenticate.cpp:102
#14 0x0000000000432b61 in CallListTask::run (this=0x7f6ae0039210) at handlers.cpp:64
#15 0x0000000000492ce6 in HttpStackUtils::SpawningHandler<CallListTask, CallListTask::Config>::process_request (
    this=<optimized out>, req=..., trail=7458)
    at /var/lib/jenkins/workspace/memento/modules/cpp-common/include/httpstack_utils.h:79
#16 0x0000000000436b6c in HttpStackUtils::HandlerThreadPool::Pool::process_work (this=<optimized out>,
    params=@0x7f6ac5fcae88: 0x7f6ac00d1d00) at /var/lib/jenkins/workspace/memento/modules/cpp-common/src/httpstack_utils.cpp:107
#17 0x0000000000437b3b in ThreadPool<HttpStackUtils::HandlerThreadPool::RequestParams*>::worker_thread_func (
    this=0x7fff381750e0) at /var/lib/jenkins/workspace/memento/modules/cpp-common/include/threadpool.h:184
#18 0x0000000000437c49 in ThreadPool<HttpStackUtils::HandlerThreadPool::RequestParams*>::static_worker_thread_func (
    pool=<optimized out>) at /var/lib/jenkins/workspace/memento/modules/cpp-common/include/threadpool.h:162
#19 0x00007f6b009cde9a in start_thread () from /lib/x86_64-linux-gnu/libpthread.so.0
#20 0x00007f6aff1c531d in clone () from /lib/x86_64-linux-gnu/libc.so.6
#21 0x0000000000000000 in ?? ()
```

**Comment by graemerobertson** [comment]
Setting to high priority since memento HTTP basically doesn't work at the moment.

**Comment by rkd-msw** [comment]
Shouldn't Monit polling spot this and restart it? Is there a separate issue here?

**Comment by bossmc** [comment]
Agreed, apparently the deadlock detection in the worker pool is not implemented in Memento/Homestead/Ralf.  This will be fixed as part of the upcoming resiliency work we're doing.

**Comment by rkd-msw** [comment]
Fixed by #252.
